### PR TITLE
Portcls fix

### PR DESCRIPTION
**Pull request by TheDeadFish**
## Purpose
portcls was handing non-contiguous buffers to drivers, this is wrong.

## Proposed changes

Split buffers on page

## TODO

Fix the race conditions?

### CODE REVIEW

**Comment by HBelusca** [comment]
What's the difference between the CIrpHandling and CIrpHandling2? Is the latter being an attempt to replace the former?

**Comment by TheDeadFish** [comment]
CIrpQueue is used in two different files and each file uses different methods. The changes I needed to make were significant and I could not easily make it work with the other methods I was not working on. So the old CIrpQueue is still needed to support the other functionality which I have not re-implemented in CIrpQueue2

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 1]
```diff
@@ -0,0 +1,597 @@
...
+#include <debug.h>
+
+class CIrpQueue2 : public IIrpQueue
+{
+public:
+    STDMETHODIMP QueryInterface( REFIID InterfaceId, PVOID* Interface);
+
+    STDMETHODIMP_(ULONG) AddRef()
+    {
+        InterlockedIncrement(&m_Ref);
```
Comment (katahiromz) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:24]: You can use the return value of increment.

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 2]
```diff
@@ -0,0 +1,597 @@
...
+        return m_Ref;
+    }
+    STDMETHODIMP_(ULONG) Release()
+    {
+        InterlockedDecrement(&m_Ref);
```
Comment (katahiromz) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:29]: You can use the return value of decrement.

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 3]
```diff
@@ -0,0 +1,597 @@
...
+
+typedef struct
+{
+    PVOID Tag;
+    UCHAR Used;
+    PVOID Data;
+    LONG Length;
+    
+    
+}KSSTREAM_TAG, *PKSSTREAM_TAG;
```
Comment (katahiromz) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:70]: Insert a space after `}`.

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 4]
```diff
@@ -0,0 +1,597 @@
...
+
+typedef struct
+{
+    ULONG nTags;
+
+    KSSTREAM_TAG Tags[];
+}KSSTREAM_DATA, *PKSSTREAM_DATA;
```
Comment (katahiromz) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:77]: Insert a space after `}`.

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 5]
```diff
@@ -0,0 +1,597 @@
...
+    volatile LONG m_Ref;
+};
+
+typedef struct
+{
+    PVOID Tag;
+    UCHAR Used;
+    PVOID Data;
+    LONG Length;
+
```
Comment (katahiromz) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:68]: Delete these two empty lines.

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 6]
```diff
@@ -0,0 +1,597 @@
...
+            return 0;
+        }
+        return m_Ref;
+    }
+    IMP_IIrpQueue;
+    CIrpQueue2(IUnknown *OuterUnknown){}
+    virtual ~CIrpQueue2(){}
+
+protected:
+
```
Comment (katahiromz) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:43]: Delete this empty line.

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 7]
```diff
@@ -0,0 +1,597 @@
...
+NTSTATUS
+NTAPI
+CIrpQueue2::QueryInterface(
+    IN  REFIID refiid,
+    OUT PVOID* Output)
+{
+    if (IsEqualGUIDAligned(refiid, IID_IUnknown))
+    {
+        *Output = PVOID(PUNKNOWN(this));
```
Comment (katahiromz) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:90]: PVOID(...) cast is not needed.

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 8]
```diff
@@ -0,0 +1,597 @@
...
+        
+        // append page count
+        StreamPageCount += ADDRESS_AND_SIZE_TO_SPAN_PAGES(
+            MmGetMdlByteOffset(Mdl), HeaderLength);
+
+        // move to next header / mdl
+        Mdl = Mdl->Next;
+        Header = (PKSSTREAM_HEADER)((ULONG_PTR)Header + Header->Size);
+
+    }while(Length);
```
Comment (katahiromz) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:197]: Insert a space after `}`.

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 9]
```diff
@@ -0,0 +1,597 @@
...
+                    (StreamData->Tags[Index].Used != FALSE))
+                {
+                    // mark mapping as released
+                    StreamData->Tags[Index].Tag = NULL;
+                    StreamData->Tags[Index].Used = FALSE;
+
+                    // done
+                    return STATUS_SUCCESS;
+                }
+
```
Comment (katahiromz) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:446]: Delete this empty line.

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 10]
```diff
@@ -0,0 +1,597 @@
...
+        }
+        
+        // append page count
+        StreamPageCount += ADDRESS_AND_SIZE_TO_SPAN_PAGES(
+            MmGetMdlByteOffset(Mdl), HeaderLength);
+
+        // move to next header / mdl
+        Mdl = Mdl->Next;
+        Header = (PKSSTREAM_HEADER)((ULONG_PTR)Header + Header->Size);
+
```
Comment (katahiromz) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:196]: Delete this empty line.

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 11]
```diff
@@ -0,0 +1,597 @@
...
+            StreamData->nTags++;
+            
+            Data = NextData;
+        }
+
+        // move to next header / mdl
+        Mdl = Mdl->Next;
+        Header = (PKSSTREAM_HEADER)((ULONG_PTR)Header + Header->Size);
+        
+    }while(Length);
```
Comment (katahiromz) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:259]: Insert a space after `}`.

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 12]
```diff
@@ -0,0 +1,597 @@
...
+
+        // move to next header / mdl
+        Mdl = Mdl->Next;
+        Header = (PKSSTREAM_HEADER)((ULONG_PTR)Header + Header->Size);
+
+    }while(Length);
+
+    // allocate stream data
+    StreamData = (PKSSTREAM_DATA)AllocateItem(NonPagedPool, sizeof(KSSTREAM_DATA) + 
+        sizeof(KSSTREAM_TAG) * StreamPageCount, TAG_PORTCLASS);
```
Comment (katahiromz) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:201]: Use a size variable to store the size.

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 13]
```diff
@@ -0,0 +1,597 @@
...
+        {
+            HeaderLength = Header->DataUsed;
+        }
+        else if (m_Descriptor->DataFlow == KSPIN_DATAFLOW_OUT)
+        {
+            HeaderLength = Header->DataUsed;
+        }
+        
+        // append pages
+        while(HeaderLength)
```
Comment (katahiromz) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:239]: Insert a space after `while`.

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 14]
```diff
@@ -0,0 +1,597 @@
...
+    Length = IoStack->Parameters.DeviceIoControl.OutputBufferLength;
+    Mdl = Irp->MdlAddress;
+    
+    do
+    {
+        /* subtract size */
+        Length -= Header->Size;    
+
+        /* get system address */
+        char* Data = (char*)MmGetSystemAddressForMdlSafe(Mdl, NormalPagePriority);
```
Comment (katahiromz) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:219]: Use LPBYTE. Declare it at the beginning of the function.

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 15]
```diff
@@ -0,0 +1,597 @@
...
+        }
+        else if (m_Descriptor->DataFlow == KSPIN_DATAFLOW_OUT)
+        {
+            HeaderLength = Header->DataUsed;
+        }
+        
+        // append pages
+        while(HeaderLength) 
+        {
+            char* NextData = (char*)ROUND_TO_PAGES(Data+1);
```
Comment (katahiromz) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:241]: Use LPBYTE. Declare it at the beginning of the function.

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 16]
```diff
@@ -0,0 +1,597 @@
...
+CIrpQueue2::CancelBuffers()
+{
+    //TODO: own cancel routine
+
+    // is there an active irp
+    if (m_Irp)
+    {
+        // re-insert it to cancelable queue
+        KsAddIrpToCancelableQueue(&m_IrpList, &m_IrpListLock, m_Irp, KsListEntryTail, NULL);
+        //set it to zero
```
Comment (katahiromz) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:317]: Insert a space after `//`.

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 17]
```diff
@@ -0,0 +1,597 @@
...
+            HeaderLength = Header->DataUsed;
+        }
+        
+        // append pages
+        while(HeaderLength) 
+        {
+            char* NextData = (char*)ROUND_TO_PAGES(Data+1);
+            LONG PageSize = NextData - Data;
+            
+            if(PageSize > HeaderLength)
```
Comment (katahiromz) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:244]: Insert a space after `if`.

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 18]
```diff
@@ -0,0 +1,597 @@
...
+    PC_ASSERT(VirtualAddress);
+    PC_ASSERT(ByteCount);
+    PC_ASSERT(Flags);
+
+    if (!m_Irp)
+    {
+        // get an irp from the queue
+        m_Irp = KsRemoveIrpFromCancelableQueue(&m_IrpList, &m_IrpListLock, KsListEntryHead, KsAcquireAndRemoveOnlySingleItem);
+        
+        if(m_Irp) {
```
Comment (katahiromz) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:350]: Insert a space after `if`.
Insert a newline after `)`.

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 19]
```diff
@@ -0,0 +1,597 @@
...
+    return STATUS_SUCCESS;
+}
+
+NTSTATUS
+NTAPI
+CIrpQueue2::GetMapping(
+    OUT PUCHAR * Buffer,
+    OUT PULONG BufferSize)
+{
+    ASSERT(0);
```
Comment (oleg-dubinskiy) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:285]:
```suggestion
    UNIMPLEMENTED;
```

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp [hunk 20]
```diff
@@ -0,0 +1,597 @@
...
+    return STATUS_NOT_IMPLEMENTED;
+}
+
+VOID
+NTAPI
+CIrpQueue2::UpdateMapping(
+    IN ULONG BytesWritten)
+{
+    ASSERT(0);
```
Comment (oleg-dubinskiy) [inline drivers/wdm/audio/backpln/portcls/irpstream2.cpp:294]:
```suggestion
    UNIMPLEMENTED;
```

**Review by oleg-dubinskiy — COMMENTED** [review]
Why not use `UNIMPLEMENTED` macros?

File: drivers/wdm/audio/backpln/portcls/irpstream2.cpp (hunk 20 above)
Comment (learn-more) [reply drivers/wdm/audio/backpln/portcls/irpstream2.cpp:294]: assert(0) is fine to see if it's being used while working on something.

Comment (oleg-dubinskiy) [reply drivers/wdm/audio/backpln/portcls/irpstream2.cpp:294]: OK.

**Comment by binarymaster** [comment]
Superseded by #3251.

**Pull request closed without merging**